lib/recursive-methods.rb

```diff
@@ -1,49 +1,101 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n == 1 || n == 0
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+
+    return s if s.length < 2
+    return reverse(s[1..-1])  + s[0]
```

Comment:
`s[1..-1]` creates a new array and  copies all the individual elements over and so is O(n) by itself.